### ISSUE TITLE
Unable to parse REP MOVSQ

### ISSUE THREAD

**Issue by sesse** [StanfordPL/stoke#853]
Hi,

It seems any function that contains REP MOVSQ (or really for me, any REP instruction) is silently ignored during extraction.

I have (for testing purposes) a binary that's simply this:

<pre>
pannekake:~/nmu/stoke> cat test.s
rep movsq
ret
pannekake:~/nmu/stoke> as test.s 
pannekake:~/nmu/stoke> ./bin/stoke extract -i a.out -o bins/
pannekake:~/nmu/stoke> grep -ri rep bins/*
pannekake:~/nmu/stoke> 
</pre>

(Repeating the test with nop instead of rep movsq works fine.)

There's some code in src/disassembler/disassembler.cc that seemingly tries to mediate between objdump's output and what x64asm wants, but I haven't managed to add a case for rep movsq that would cause the function to work properly.

**Comment by bchurchill** [comment]
That's right -- we don't support REP-prefixed instructions.  I don't think it's even in x64asm.

**Comment by bchurchill** [comment]
If you or anyone else is interesting in working on this, I don't think it would be too bad.  It looks like some support exists in x64asm -- probably the only thing to do is to improve the parser (and the src/disassembler/disassembler.cc code) to accept the rep prefix.  The harder part would be making the sandbox in src/sandbox/sandbox.cc work.  The sandbox is essentially a very lightweight x86-64 virtual machine that runs instructions natively.  Before we run an instruction, we need to sanitize its arguments to make sure it doesn't read/write the wrong places.  For rep instructions, we haven't worked this out yet, but it shouldn't be too hard to do.

**Comment by sesse** [comment]
OK, so this means essentially anything doing memcpy() is right out? (My function uses it, and I got an error that _memcpy wasn't defined, so I compiled again with -minline-all-stringops, which of course spits out the rep prefix :-) )

**Comment by bchurchill** [comment]
You should be able to use memcpy().  The reason it's not defined is
because it's dynamically linked with libc when your binary is run.  The
easiest thing to do is to implement a memcpy-like stub in a separate .s
file.  Then, when you run stoke, use --functions to specify a folder
containing this file, along with any others that you need.  Here's a
version from gcc compiled with -Os.  -Os will give us something as small
as possible, which is generally good for STOKE.  I haven't tested this,
but try the following memcpy.s file:

```
  .text
  .globl memcpy
  .type memcpy, @function

#! file-offset 0x526
#! rip-offset  0x400526
#! capacity    26 bytes

# Text                      #  Line  RIP       Bytes  Opcode           
.memcpy:                    #        0x400526  0      OPC=<label>      
  movq %rdi, %rax           #  1     0x400526  3      OPC=movq_r64_r64 
  movl %edx, %edx           #  2     0x400529  2      OPC=movl_r32_r32 
  xorl %ecx, %ecx           #  3     0x40052b  2      OPC=xorl_r32_r32 
.L_40052d:                  #        0x40052d  0      OPC=<label>      
  cmpq %rdx, %rcx           #  4     0x40052d  3      OPC=cmpq_r64_r64 
  je .L_40053f              #  5     0x400530  2      OPC=je_label     
  movb (%rsi,%rcx,1), %dil  #  6     0x400532  4      OPC=movb_r8_m8   
  movb %dil, (%rax,%rcx,1)  #  7     0x400536  4      OPC=movb_m8_r8   
  incq %rcx                 #  8     0x40053a  3      OPC=incq_r64     
  jmpq .L_40052d            #  9     0x40053d  2      OPC=jmpq_label   
.L_40053f:                  #        0x40053f  0      OPC=<label>      
  retq                      #  10    0x40053f  1      OPC=retq         

.size memcpy, .-memcpy
```
